### DISCUSSION TITLE
Adapting python components from the Urban Weather Generator into SOLWEIG

### DISCUSSION THREAD

**Discussion by rarygit** [UMEP-dev/UMEP#235] [Ideas]
The Urban Weather Generator (UWG) was translated from Matlab (Bueno, Nakano, Yang) to Python by Saeran Vasanthakumar and uploaded to GitHub.

UWG can be run as a standalone python application from the CLI, in conjunction with the uwg-schema package.
To run as such, it does not require any components from the rhino plugins (e.g. grasshopper etc).

I am curious to know whether components of the UWG model can be of use for SOLWEIG?
For example the turbulence model?

There is a case study for the City of Vienna, which compares UWG with the WRF model.
https://link.springer.com/content/pdf/10.1007%2Fs12273-019-0564-y.pdf

"The results further indicated that the meteorological output from the UWG and morphing models is closer to the monitored data, when compared to the output from the WRF model."

https://github.com/ladybug-tools/uwg
https://github.com/ladybug-tools/uwg/blob/6ff8e44ef0a35217d9f9bfc213c4dcbb926fdeff/uwg/UCMDef.py

https://github.com/ladybug-tools/uwg-schema

API docs:  https://www.ladybug.tools/uwg/docs/uwg.cli.html

**Comment by biglimp** [comment]
I know about the work of Bueno but I didnt know that there was a Python version available. Potentially, very interesting! Regarding turbulence (wind), UWG is a local scale model whereas SOlWEIG is solving radiation at the micro scale. However, using UWG to force SOLWEIG is definelty something to look into.

**Reply by rarygit** [reply]
I installed uwg in a new conda environment this morning. It passed all tests and good to go.
I sent you an email - did that arrive?

These are the commands I used to install UWG on python:

(i) Create uwg environment

conda create --name uwg1

(ii) Install git and pip

conda install git
conda install pip

(iii) Local Development - Clone this repo locally

git clone https://github.com/ladybug-tools/uwg

(iv) Install dependencies:

cd uwg
pip install -r dev-requirements.txt
pip install -r requirements.txt
python3 setup.py install

(v) Run Tests:

python -m pytest tests/

(vi) Generate Documentation:

sphinx-apidoc -f -e -d 4 -o ./docs ./uwg
sphinx-build -b html ./docs ./docs/_build/docs

**Comment by biglimp** [comment]
What OS are you using? One crucial point that has to be solved is if the tool will function within QGIS. One easy test is to execute the tool via  Python console in QGIS. Also, the dependencies used by UWG should not be in conflict with the QGIS software system.

**Reply by rarygit** [reply]
I am using Ubuntu 20.04

QGIS v3.14 "Pi" is installed in conda env qgis_pi
UWG is installed in conda env uwg1, as above.

Now I will install uwg in env qgis_pi and see what happens.

**Reply by rarygit** [reply]
You also have to run "python3 setup.py install" in the uwg folder.
I edited the post regarding that step.
uwg was missing from the "conda list" output ....

Attached is the conda install log and the list of packages I have in the conda "QGIS v3.14 Pi" environment.

There were no conflicts (I use conda-forge channel only; not anaconda channel).

I needed to install 'decorator' separately (conda install decorator), as it was missing.

Now I will try running UWG in the QGIS python console. 

[conda uwg install log_and list of conda packages.txt](https://github.com/UMEP-dev/UMEP/files/5971401/conda.uwg.install.log_and.list.of.conda.packages.txt)

**Reply by rarygit** [reply]
There were no dependency conflicts with QGIS.

In running the example script from the QGIS python console, there was an "Assertion Error", which I will have to interpret:
AssertionError: zone must be on of ('1A', '2A', '2B', '3A', '3B-CA', '3B', '3C', '4A', '4B', '4C', '5A', '5B', '6A', '6B', '7', '8'). Got: 1.

Beginner's learning curve ......
Other than that, everything worked OK.

++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
From the UWG GitHub page, there is this example (https://github.com/ladybug-tools/uwg):

Here is a Python example that shows how to create and run an Urban Weather Generator object.

`from uwg import UWG`
`# Define the .epw, .uwg paths to create an uwg object.`
`epw_path = "SGP_Singapore.486980_IWEC.epw"`
` # Initialize the UWG model by passing parameters as arguments, or relying on defaults`
`model = UWG.from_param_args(bldheight=10, blddensity=0.5, vertohor=0.8, grasscover=0.1, 
                            treecover=0.1, zone='1A')`
`# Uncomment these lines to initialize the UWG model using a .uwg parameter file`
`# param_path = "initialize_singapore.uwg"  # available in the resources directory.`
`# model = UWG.from_param_file(param_path, epw_path=epw_path)`
`model.generate()`
`model.simulate()`
`# Write the simulation result to a file.`
`model.write_epw()`

+++++++++++++++++++++++++++++++++++++++++++++++++++

In the QGIS python console, this is how it went:

Python Console
Use iface to access QGIS API interface or Type help(iface) for more info
Security warning: typing commands from an untrusted source can lead to data loss and/or leak

epw_path = "/home/l1nux/miniconda3/envs/uwg1/uwg/tests/epw/SGP_Singapore.486980_IWEC.epw"
param_path = "/home/l1nux/miniconda3/envs/uwg1/uwg/resources/initialize_singapore.uwg"
model = UWG.from_param_file(param_path, epw_path=epw_path)

Traceback (most recent call last):
  File "/home/l1nux/miniconda3/envs/qgis_pi/lib/python3.8/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
  File "/home/l1nux/miniconda3/envs/qgis_pi/lib/python3.8/site-packages/uwg-5.8.4.dev7+gc0fac9f-py3.8.egg/uwg/uwg.py", line 246, in from_param_file
    model._read_input(param_path)
  File "/home/l1nux/miniconda3/envs/qgis_pi/lib/python3.8/site-packages/uwg-5.8.4.dev7+gc0fac9f-py3.8.egg/uwg/uwg.py", line 1487, in _read_input
    setattr(self, attr, self._init_param_dict[attr])
  File "/home/l1nux/miniconda3/envs/qgis_pi/lib/python3.8/site-packages/uwg-5.8.4.dev7+gc0fac9f-py3.8.egg/uwg/uwg.py", line 854, in zone
    assert value in REF_ZONETYPE_SET, 'zone must be on of {}. Got: {}.'.format(
AssertionError: zone must be on of ('1A', '2A', '2B', '3A', '3B-CA', '3B', '3C', '4A', '4B', '4C', '5A', '5B', '6A', '6B', '7', '8'). Got: 1.

**Reply by rarygit** [reply]
Ok, I think this line should be:

`model = UWG.from_param_args(bldheight=10, blddensity=0.5, vertohor=0.8, grasscover=0.1, treecover=0.1, zone='1A',)`

There should be a comma after '1A' and before the parentheses:  zone='1A',)   and not:    zone='1A')

**Reply by rarygit** [reply]
Checking the issues section on the UWG GitHub page, there seems to have been errors running this example for other users: 
[https://github.com/ladybug-tools/uwg/issues/139](url)

**Reply by biglimp** [reply]
Thanks for this report. From a user-based perspective, I also wonder how this will work in a W10 environment...

**Reply by rarygit** [reply]
Happy to say I managed to get UWG to run OK in the QGIS Python Console -- it passes the crucial test.
UWG runs inside QGIS.

I downloaded another .epw file (ARE_Abu.Dhabi.412170_IWEC.epw) from here:  [https://energyplus.net/weather](url)
And changed its name to:    ARE_AbuDhabi_412170_IWEC.epw 
And made sure that the epw_path was formatted correctly.

Now that works, what is the next step?

+++++++++++++++++++++++++++++++++++++++++++
Python Console
Use iface to access QGIS API interface or Type help(iface) for more info
Security warning: typing commands from an untrusted source can lead to data loss and/or leak
from uwg import UWG
epw_path = '/home/l1nux/Downloads/ARE_AbuDhabi_412170_IWEC.epw'
bld = [('largeoffice', 'pst80', 0.4), ('midriseapartment', 'pst80', 0.6)] 
model= UWG.from_param_args(bldheight=10, blddensity=0.5, vertohor=0.8, grasscover=0.1, treecover=0.1, zone='1A', bld=bld, epw_path=epw_path)
model.generate()
model.simulate()
Simulating new temperature and humidity values for 31 days from 1/1.
Simulating Day 1
Simulating Day 2
Simulating Day 3
Simulating Day 4
Simulating Day 5
Simulating Day 6
Simulating Day 7
Simulating Day 8
Simulating Day 9
Simulating Day 10
Simulating Day 11
Simulating Day 12
Simulating Day 13
Simulating Day 14
Simulating Day 15
Simulating Day 16
Simulating Day 17
Simulating Day 18
Simulating Day 19
Simulating Day 20
Simulating Day 21
Simulating Day 22
Simulating Day 23
Simulating Day 24
Simulating Day 25
Simulating Day 26
Simulating Day 27
Simulating Day 28
Simulating Day 29
Simulating Day 30
Simulating Day 31
model.write_epw()
New climate file is generated at /home/l1nux/Downloads/ARE_AbuDhabi_412170_IWEC_UWG.epw.

**Reply by biglimp** [reply]
Nice! Can you show a time series plot where you compare the forcing air temperature with the the modeled one. 
One question is how flexible it is to use other input forcing data, e.g. a UMEP-formatted file? Also, the Windows issue should to the checked (see above). It would be verg good if you can get it to work with ERA5 data also...

To put in a tool in UMEP would be rather straight forward. As far as I see, all parameters can be derived from tools already present. Maybe some thought needs to go in regarding the bld variable...

**Reply by rarygit** [reply]
Regarding W10:

I would install miniconda3 (not anaconda), 
This is an importent document to read: Managing Conda environments
[https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html](url)

at the prompt create a conda environment (conda --name qgis_pi), 
make sure you set conda-forge as the main channel (conda config --add channels conda-forge);
then install QGIS (conda install qgis);
and finish with the steps in conda for UWG as I outlined above.

Let me know if there are problems.

**Reply by biglimp** [reply]
We should not use anaconda but instead use what is already in QGIS and the OSGeo installation.

**Reply by rarygit** [reply]
Can you show a time series plot where you compare the forcing air temperature with the the modeled one?

Yes, I will do this. I will take a short pause because getting the example to run correctly cost a few neurons .....

**Reply by rarygit** [reply]
Yes, regarding OSGEO - 

I never had problems with python dependencies using miniconda (e.g. netcdf); but i did with OSGEO

For me, conda is almost failsafe as I can simply delete each conda environment if and when they have problems.

But OSGEO is fine as the rule, so there is no misunderstanding.

But once you know conda env, you never go back ... ;- )

**Reply by biglimp** [reply]
I know, but I always tend to think from a user-perspective and the fewer steps to reach a result, the better. Imagine a person installing QGIS and UMEP and then he/she is ready to use the tools available without any issue of additional installations and missing python library dependencies etc...

**Reply by rarygit** [reply]
I agree with you to use OSGEO as the default, because there is only one installation interface to manage when supporting the user base.

For those who can work with conda, such as myself, this is something extra. 

This is more a development and testing interface - as we can see here in relation to determining the interoperability between QGIS and UWG.  

When i was using W8, W10, it became an issue to develop in Python outside the QGIS python environment. 
So I started to use conda env to manage those issues, without encroaching on the environments and paths set by the WIndows OS itself.

**Reply by biglimp** [reply]
I have been using W8 and W10 for all my Python stuff using the OSGeo installation. I know all about it ;-) so hopefully we can get it to work.

**Reply by rarygit** [reply]
Here is the information regarding weather data files.

As I understand it, UWG uses the epw format, which is an ascii csv format conforming to the EnergyPlus application.
There is bound to be a python script for this, somewhere.

In this regard, you could email the main python developer for uwg:   Saeran Vasanthakumar,  saeranv(at)gmail(dot)com
GitHub page here:  [https://github.com/saeranv](url)
 
There is a free file converter from csv to epw called weather converter inside EnergyPlus on GitHub:
[https://github.com/NREL/EnergyPlus/tree/develop/bin/WeatherConverter](url)

There is also the open source free software called Elements, available here:
[https://bigladdersoftware.com/projects/elements/](url)
This offers copy and paste from spreadsheets into Elements, which saves the epw file.
There are some other utilities in Elements, such as making charts from epw files.
I will try this to make the chart you asked for.

The headers for epw files are not complicated and I am sure there will be an easy solution, based on csv or txt files, to do the necessary conversions.

Some Literature on the file conversion process:
[https://bigladdersoftware.com/epx/docs/8-3/auxiliary-programs/using-the-weather-converter.html#select-file-to-convert](url)
[https://bigladdersoftware.com/projects/elements/docs/user-guide/](url)
[https://energyplus.net/sites/default/files/pdfs_v8.3.0/Tips_and_Tricks_Using_EnergyPlus.pdf](url)

As well as the LadyBug Tools forum (Rhino plugin):
[https://discourse.ladybug.tools/search?q=csv](url)    -- search term "csv"
 
And the UWG python forum and GitHub UWG issues page.

**Reply by rarygit** [reply]
I transferred the input and output epw data to ods and xlsx format using the free Elements program.

Without studying the simulation, and remembering that I replaced the Singapore site with Abu Dhabi, there is very little difference, if any, between the dry bulb temperatures.



[ARE_AbuDhabi_412170_IWEC_UWG_epw.xlsx](https://github.com/UMEP-dev/UMEP/files/5972956/ARE_AbuDhabi_412170_IWEC_UWG_epw.xlsx)

![Chart_Output_Wet DryBulbTemperatures](https://user-images.githubusercontent.com/16574504/107791815-a6ea8580-6d54-11eb-87ca-57d720abe8d7.png)
![Chart_Output_WindSpeed_WindDirection](https://user-images.githubusercontent.com/16574504/107791818-a81bb280-6d54-11eb-9535-08dfc523c721.png)
![Elements_Output_Add_Variables](https://user-images.githubusercontent.com/16574504/107791820-a81bb280-6d54-11eb-9eb5-76d16f7a5e63.png)
![Elements_Output_Some_Variables](https://user-images.githubusercontent.com/16574504/107791822-a8b44900-6d54-11eb-86ad-137c0215a72a.png)

**Reply by biglimp** [reply]
Seems to produce some kind of urban effect. Would be good to know what parameters in the epw-file that is required. Translating from an UMEP formatted metfile to epw should not be an issue.We have a translation from epw to UMEP available already. This issue would be if there are some variables used in the UWG that is not commonly available in UMEP. In the long run it would also be possible to use ERA5 data.

![image](https://user-images.githubusercontent.com/11908356/107795668-427df500-6d59-11eb-9970-c18a7a07d402.png)

**Reply by rarygit** [reply]
A graphical overview of the parameter workflow is here:
[http://hydrashare.github.io/hydra/viewer?owner=chriswmackey&fork=hydra_2&id=Urban_Weather_From_Parameters&slide=0&scale=1.840375301249751&offset=-147.3137770718776,-217.59536759804502](url)

And here, Workflows for Urban Weather Generator Input:
[https://github.com/ladybug-tools/dragonfly-legacy/wiki/1.-UWG-Workflows](url)

++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

The MIT website for the UWG has available the MSc theses (Nakano and Yang) and the PhD dissertation (Bueno), which provide quite good theoretical material. Especially the PhD thesis. 

The article by Vuckovic et al. 2020 is also helpful.

Finally the API documentation for uwg and uwg_schema.

**Reply by biglimp** [reply]
This one? Vuckovic, M., Tötzer, T., & Stollnberger, R., & Loibl, W. (2020). Urban transformation and heat island: Potential of urban design alternatives to mitigate the effects of urban overheating in Austrian cities. J. Urban Env. 01(01), 03 – 14.  

The link provide no info on what met variables that is actually required. What is the link to the MIT/UWG website?

**Reply by rarygit** [reply]
To save you time, I have shared a dropbox folder with you that contains the results I have from a desktop search. 

(A) Taking UWG as a whole  - connecting EnergyPlus, building BIM, vegetation canopy:

From memory, I understand that Bruno's method involved adapting a DOE epw file adjacent to your area of interest. Then morphing (sic.) the air temperature according to the expected UHI influencing your built/vegetated urban area of interest. 
To that extent 
(i) the meteorological variables are consistent with the epw file;
(ii) the building variables (BIM), which I haven't looked at in detail;
(iii) same for the vegetation, soil and irrigation/water inputs.

(B) My initial thought (a smaller step) was whether turbulence (wind direction, wind speed), any of the vegetation/canopy python algorithms and the UWG vertical discretisation could be incorporated into SOLWEIG.

**Reply by rarygit** [reply]
As the python version of UWG was extensively reworked by Chris Mackey as a Rhino CAD plugin (building BIM and grid layout), most of the material describing the workflow comes from there.

[https://discourse.ladybug.tools/t/dragonfly-0-0-3-and-urban-weather-generator-5-0-released/3250](url)

[https://grasshopperdocs.com/components/dragonfly/runUrbanWeatherGenerator.html](url)

[https://www.ladybug.tools/resource.html](url)

Outside of this workflow, you can run UWG from the Python console. However you have to go back to the Theses and Dissertation to ascertain the driving variables relative to an UMEP-specific implementation. While keeping an eye on Chris Mackey's workflows for UWG in Rhino.

[http://web.mit.edu/SustainableDesignLab/publications/BS2015_UWG_KendallSquare.pdf](url)  MIT article;
[http://urbanmicroclimate.scripts.mit.edu/](url)  MIT links;

UWG parameters here (MIT, matlab)  [http://urbanmicroclimate.scripts.mit.edu/uwg_parameters.php](url)

xml file - provides the parameterisation (matlab version)  [http://urbanmicroclimate.scripts.mit.edu/files/sample_inputs.xml?v=0.1](url)


[sample_inputs.xml.txt](https://github.com/UMEP-dev/UMEP/files/5980931/sample_inputs.xml.txt)

.

**Reply by rarygit** [reply]
Just from scanning through these two links, I can see useful input data overlaps between UWG, SUEWS and SOLWEIG.

UWG parameters here (MIT, matlab) http://urbanmicroclimate.scripts.mit.edu/uwg_parameters.php

xml file - sample input file (matlab version) http://urbanmicroclimate.scripts.mit.edu/files/sample_inputs.xml?v=0.1

**Reply by biglimp** [reply]
Thanks for all info. THe last links are only building and surface related and those should be no problem to derive from other UMEP tools. I am concerned by what meteorological variables that is needed in the UWG. There are a lot that is not used (probably). My main concern is if UWG needs e.g. diffuse and direct solar radiation, incoming longwave, wind speed and direction, total sky cover etc... I assume it needs the common ones such as air temperature, relative humidity, global radiation, pressure... I will have a look when I find the time.

**Reply by rarygit** [reply]
Additional Information.

EPW file; weather data dictionary here:
https://bigladdersoftware.com/epx/docs/8-3/auxiliary-programs/energyplus-weather-file-epw-data-dictionary.html#energyplus-weather-file-epw-data-dictionary

EnergyPlus Input requirements (description of weather data/epw file):
https://energyplus.net/sites/all/modules/custom/nrel_custom/pdfs/pdfs_v9.4.0/InputOutputReference.pdf

R package:  epwshiftr
https://github.com/ideas-lab-nus/epwshiftr
Create Future 'EnergyPlus' Weather Files using 'CMIP6' Data
https://github.com/ideas-lab-nus

Some of the LadyBug Tools on GitHub are standalone python toolsets.
For example LadyBug Geometry.
These standalone python toolsets do not require the proprietary CAD program, Rhino.
https://github.com/ladybug-tools/

**Reply by rarygit** [reply]
When I needed rather obscure meteorological input for Tethys-Urban (SAB1, SAD1, see Output below),  Simone Fatichi recommended using his AWE-GEN application (Advanced WEather GENerator) coded in Matlab:
https://hyd.ifu.ethz.ch/research-data-models/awe-gen.html

This would be useful when converted to Python and included in UMEP.

The AWE-GEN model can take existing meteorological data, as well as simulate new time series when the user has no data but provides a file containing the parameter set of the model.  

Output:
Prs    Precipitation [mm h -1 ]
Ns     Cloud cover [0-1]
Tas    Air temperature [°C]
Rsws   Global solar radiation [W m -2 ]
Us     Relative Humidity [0-1]
Wss    Wind Speed [m s -1 ]
Pres   Atmospheric Pressure [mbar]
Rdirs  Direct solar radiation [W m -2 ]
Rdifs  Diffuse solar radiation [W m -2 ]
SAB1   Direct solar radiation UV/VIS band [W m -2 ]
SAD1   Diffuse solar radiation UV/VIS band [W m -2 ]
SAB2   Direct solar radiation NIR band [W m -2 ]
SAD2   Diffuse solar radiation NIR band [W m -2 ]
PARB   Direct photosyntetically active solar radiation [W m -2 ]
PARD   Diffuse photosyntetically active solar radiation [W m -2 ]

Date:  This is 4 vectors Yrs [Years] Mos [Months] Das [Days] Hrs [Hours] and one vector DS in matlab date format.

**Reply by biglimp** [reply]
This does not answer my question. What parameters in a epw-file is required bu UWG?

**Reply by rarygit** [reply]
I'm not an expert with uwg, I'm learning as we look at it together (in between work).
See Parameters below
My question, is it feasible for SOLWEIG or even UMEP?

https://www.ladybug.tools/uwg/docs/uwg.weather.html

uwg.weather module
Read EPW file[1] and store weather timeseries data.

class uwg.weather.Weather(epw_path, HI, HF)

climate_file – Text string for the name of the rural epw file that will be morphed.
HI – Number for initial EPW row based on final Julian data.
HF – Number for final EPW row based on start Julian data.

Parameters:
staTemp – drybulb (K)
staTdp – dewpoint (C)
staRhum – air RH (%)
staPres – air pressure (Pa)
staInfra – horizontal Infrared Radiation Intensity (W m-2)
staHor – horizontal radiation [W m-2]
staDir – normal solar direct radiation (W m-2)
staDif – horizontal solar diffuse radiation (W m-2)
staUdir – wind direction
staUmod – wind speed (m s-1)
staRobs – precipitation (mm h-1)
staHum – specific humidty (kgH20 kgN202-1)

**Reply by biglimp** [reply]
Yes, Definitely feasible.
Perfect, Just what I am looking for regarding weather parameters. Surprised that wind direction is needed. All parameters already including in standard UMEP met format.

**Reply by rarygit** [reply]
I think wind direction, along with wind speed, is needed to calculate the canyon turbulent velocities:
https://www.ladybug.tools/uwg/docs/uwg.UCMDef.html

turbU – canyon turbulent velocities (m s-1)
turbV – canyon turbulent velocities (m s-1)
turbW – canyon turbulent velocities (m s-1)

Canyon turbulence is of particular interest to me; regarding the total cooling factor as a function of shading, soil moisture and vegetation canopy shape. This is why I asked about incorporating UWG, to get access to this feature.

And SOLWEIG lacks this? Subsequently the UTCI results are at each POI? 
I think this was the answer you gave to a previous question.

**Comment by biglimp** [comment]
Wind direction is included in UMEP. The thing is that UWG does not include canyon alignment (direction) as far as I can see, so wind direction becomes trivial. 

You should remember that UWG generator is a local scale model whereas SOLWEIG is a micro scale. UWG would be incorporated into UMEP to feed e.g. SOLWEIG with local scale air temperature and humidity. Micro scale pedestrian wind (canyon turbulance) will not be solved by including UWG into UMEP.

**Reply by rarygit** [reply]
Yes that is true in relation to this version of the UWG, which is covered in some detail in Bueno's dissertation regarding wind velocity and the magnitude of the UHI effect.

And it has been good to work through together in this discussion whether UWG is feasible for SOLWEIG/UMEP. 

What I also found at the same time was a modification of the UWG, which predicts vertical profiles of turbulence kinetic energy. As well as the usual outputs generated by UWG, including urban building energy performance.  

This is the urban microclimate application Vertical City Weather Generator (VCWG v1.3, 2019, Moradi & Aliabadi) coded in Python2.
https://zenodo.org/record/425522
The Zenodo record has all versions, 1.0 - 1.3. And v1.0 release is also on GitHub.
https://github.com/MMoradi-Eng/VCWG

I have installed and run VCWG in a conda env for python 2 on Ubuntu 20.04.

I have attached seven python files that are not in UWG, which are associated with the VCWG's TKE calculations, numerical solver and a lightweight tree energy balance. What do you think? Perhaps we also need to consider the GMD reviewer's comments regarding some of the assumptions in the UWG model. See below.

It needs to be said that the authors preprint paper, based on VCWG, was "robustly" reviewed in the discussion on GMD.  Including strong criticism of UWG's rural model. Accordingly the VCWG author's made significant changes.
https://gmd.copernicus.org/preprints/gmd-2019-176/gmd-2019-176-AC1-supplement.pdf
A revised version of the preprint was accepted for the journal GMD and is expected to appear there eventually.
https://gmd.copernicus.org/preprints/gmd-2019-176/ 
Authors: Mohsen Moradi, Benjamin Dyer, Amir Nazem, Manoj K. Nambiar, M. Rafsan Nahian, Bruno Bueno, Chris Mackey, Saeran Vasanthakumar, Negin Nazarian, E. Scott Krayenhoff, Leslie K. Norford, and Amir A. Aliabadi

Links for Amir Aliabadi, VCWG co-author and Assoc. Prof. UOGuelph:
http://www.aaa-scientists.com/
http://www.aaa-scientists.com/Aliabadi%202018.pdf
https://www.researchgate.net/profile/Amir_A_Aliabadi

[Buoyancy.py.txt](https://github.com/UMEP-dev/UMEP/files/5989927/Buoyancy.py.txt)
[ColModel.py.txt](https://github.com/UMEP-dev/UMEP/files/5989928/ColModel.py.txt)
[DragLength.py.txt](https://github.com/UMEP-dev/UMEP/files/5989929/DragLength.py.txt)
[DragTurb.py.txt](https://github.com/UMEP-dev/UMEP/files/5989930/DragTurb.py.txt)
[NumericalSolver.py.txt](https://github.com/UMEP-dev/UMEP/files/5989931/NumericalSolver.py.txt)
[Shear.py.txt](https://github.com/UMEP-dev/UMEP/files/5989932/Shear.py.txt)
[Tree_EnergyBalance.py.txt](https://github.com/UMEP-dev/UMEP/files/5989933/Tree_EnergyBalance.py.txt)

**Comment by rarygit** [comment]
Hi Csilla, I didn't see your email until now. I was corresponding in the Discussion section of UMEP-Dev and in the last post suggested VCWG.The VCWG co-developer Dr. Aliabadi (AOGuelph) replied to me yesterday that he and Dr. Mohani would be available for answering questions to integrate VCWG at the microscale.By that I meant a raster grid where 1 pixel = 1 m².Dr. Aliabadi also wrote that the final copy of their GMD article (linked in my Discussion post) will be put on line in the next two to three days. Best regards,Tim 16.02.2021, 17:50, "Csilla V Gal" <notifications@github.com>: Hi Fredrik,As you know, UWG only "urbanizes" air temperature and relative humidity (while leaving dew point temperature values untouched). It requires the weather file to be in epw format. I could check later which are the minimum weather parameters that that the program will still runs with. However, if you need a quick answer, it's better if you reach out to the team https://pypi.org/project/uwg/For wind, there is VCWG (https://zenodo.org/record/3376792). It is a sort of extension of UWG that predicts the vertical profile of wind speed as well. According to some prelim. results, it considers canyon orientation in radiation, but I would need to test for wind speed/direction. It too is in Python (v2.7). However, it is still in development and hence, a bit buggy and the latest changes/developments lack documentation.Best,Csilla—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

**Reply by CsillaVG** [reply]
Hi rarygit,

I deleted the my initial comment since I noticed that you posted yours with very similar content just a minute or so before mine. I too was planning to reach out to the authors (when my time allows), but it's great that you did.
I also see that there is a newer version out now (I tested up to v1.3.0). I will check those out too.
Thank you for the many info and update!
Best,
Csilla

**Reply by rarygit** [reply]
Hi Csilla,

Thanks for joining the discussion, your comments were important.
And please continue, it's helpful.

This morning I set up a conda environment for python 2.7.13 on Ubuntu 20.04 and completed the test simulation for VCWG v1.3.2.  As I said to Amir and Mohsen, I couldn't test VCWG with QGIS 2. There are too many conflicts. In the attached file, you can see that matplotlib is installed in the conda env using "pip install". Too many conflicts if you use "conda install".

Mohsen wrote that it is possible to apply the model within a raster file as long as you have the spatial distribution of the input parameters. I would say that SOLWEIG/UMEP can provide that, yes? Mohsen is developing another code to automate this process for any given region. Both are interested in the concept of a QGIS plugin. I have reiterated the importance, well at least for me, that this bespoke turbulence feature can operate at the microscale.

I asked Amir and Mohsen whether the seven python scripts I attached above could be converted to python 3. 
As well as any tweaks they made to the VCWG v1.3.2 model version of UWG.
This 'might' present a relatively quick path to bring VCWG into the existing UWG python 3 framework, available on GitHub

I ran the VTUF-3D model example yesterday, the PrestonBase8. That consumed 6.4 Gb of space and took quite a while. MAESPA 3D provides the opportunity to change the tree canopy shape. In Tethys-Urban you cannot, evidently the re-coding would be arduous. But I am 'hoping' that the VCWG/SOLWEIG/UMEP development can reduce the output size and computation time, while providing further interesting functionality at the microscale. 

[Successful_Run.log](https://github.com/UMEP-dev/UMEP/files/6001534/Successful_Run.log)

**Reply by rarygit** [reply]
Amir has emailed the final version of the VCWG paper and supplement that was just published on GMD.


[Moradi et al. 2021 The Vertical City Weather Generator (VCWG v1.3.2).pdf](https://github.com/UMEP-dev/UMEP/files/6003390/Moradi.et.al.2021.The.Vertical.City.Weather.Generator.VCWG.v1.3.2.pdf)
[Moradi et al. 2021 The Vertical City Weather Generator (VCWG v1.3.2) Supplement.pdf](https://github.com/UMEP-dev/UMEP/files/6003391/Moradi.et.al.2021.The.Vertical.City.Weather.Generator.VCWG.v1.3.2.Supplement.pdf)

**Reply by CsillaVG** [reply]
Great! Thank you.

**Comment by rarygit** [comment]
I converted VCWG v1.3.2 files from python 2 to python 3.

It can now run on python version 3.9 (conda env).

Next will try to get it to work from the QGIS python console.

**Reply by CsillaVG** [reply]
Hi Rarygit,
Quick question: Do you know by any chance which is VCWG authors' preferred way of communicating bugs and other issues?
Thanks!

**Comment by rarygit** [comment]
Hi Csilla, that is a good question. 

I will ask Mohsen if he could upload/update his GitHub VCWG repository to version 1.3.2, so that issues and discussions can take place there. You could also email Mohsen directly, his address is in the article.

I have attached the Python 3 version of VCWG 1.3.2 if you are interested. 
I used 2to3.py in conda to step through the conversion while I ran VCWG. It is a rough-cut.

Maybe we need to create a pypi package, already done for uwg, so that the VCWG modules be installed in a conda/QGIS environment? 

[VCWG v1-3-2_(Python3-8_QGIS3-16).tar.gz](https://github.com/UMEP-dev/UMEP/files/6140109/VCWG.v1-3-2_.Python3-8_QGIS3-16.tar.gz)
